File: python/grasp.py
from neighborhoods import *
from local_search import StoppingCriterion, MaxIterations, StepFunction, local_search
from typing import Callable, Optional
from construction import calc_my_metric
import random
        

# def randomized_constructor(
#     I: Instance,
#     alpha: float = 0.3
# ) -> Solution:

#     sol = Solution(routes=[[] for _ in range(I.nK)])

#     # --- score each request by direct pickup->drop distance ---
#     scores = np.zeros(I.n)
#     for r in range(I.n):
#         pu = 1 + r
#         dr = 1 + I.n + r
#         scores[r] = I.dist[pu, dr]  # simple greedy metric

#     # --- sort requests by greedy score ---
#     req_order = np.argsort(scores)

#     served = 0
#     used = set()

#     for _ in range(I.n):

#         # remaining requests
#         remaining = [r for r in req_order if r not in used]
#         if not remaining:
#             break

#         # greedy interval
#         k = max(1, int(alpha * len(remaining)))
#         rcl = remaining[:k]          # restricted candidate list
#         req = random.choice(rcl)     # pick random request from RCL
#         used.add(req)

#         pickup = 1 + req
#         drop = 1 + I.n + req
#         demand = I.demands[req]

#         best_route = None
#         best_pos = None
#         best_delta = math.inf

#         # --- try all vehicles and insertion positions ---
#         for vk in range(I.nK):
#             route = sol.routes[vk]

#             m = len(route)
#             for ip in range(m + 1):
#                 for jp in range(ip + 1, m + 2):

#                     # candidate route
#                     new_route = route[:ip] + [pickup] + route[ip:jp] + [drop] + route[jp:]

#                     cargo = calc_route_cargo(I, new_route)
#                     if np.any(cargo > I.C) or np.any(cargo < 0):
#                         continue  # infeasible

#                     d_old = route_distance(I, route)
#                     d_new = route_distance(I, new_route)
#                     delta = d_new - d_old

#                     if delta < best_delta:
#                         best_delta = delta
#                         best_route = vk
#                         best_pos = (ip, jp)

#         if best_route is None:
#             # cannot insert further; stop early
#             break

#         # --- apply best insertion ---
#         ip, jp = best_pos
#         r = sol.routes[best_route]
#         sol.routes[best_route] = r[:ip] + [pickup] + r[ip:jp] + [drop] + r[jp:]
#         served += 1

#         if served >= I.gamma:
#             break

#     return sol

def randomized_constructor_simple(
    I: Instance,
    a: float = 0.5,
    alpha: float = 0.3,
    max_tries: int = 20
) -> Solution:

    costs = calc_my_metric(I, a)

    # sort requests by heuristic cost
    perm = np.argsort(costs)
    n = I.n
    
    sol = Solution(routes=[[] for _ in range(I.nK)])
    served = 0
    used = set()

    while served < I.gamma:

        # --- RCL creation ---
        remaining = [r for r in perm if r not in used]
        if not remaining:
            break

        k = max(1, int(alpha * len(remaining)))
        rcl = remaining[:k]

        req = random.choice(rcl)
        used.add(req)

        pickup = 1 + req
        drop = 1 + I.n + req

        inserted = False

        # --- Try random routes and random insertion positions ---
        for _ in range(max_tries):

            vk = random.randrange(I.nK)
            route = sol.routes[vk]
            m = len(route)

            # generate random insertion points
            ip = random.randint(0, m)
            jp = random.randint(ip + 1, m + 1)

            new_r = route[:ip] + [pickup] + route[ip:jp] + [drop] + route[jp:]
            cargo = calc_route_cargo(I, new_r)

            if np.all((0 <= cargo) & (cargo <= I.C)):
                sol.routes[vk] = new_r
                inserted = True
                served += 1
                break

        # if failed to insert this request, skip it and continue
        if not inserted:
            continue

    return sol

        
def grasp(
    I: Instance,
    randomized_constructor: Callable[[Instance], Solution],
    neighborhoods_cls: List[type[Neighborhood]],
    step_function: StepFunction.Func_t,
    stopping_criterion_outer: StoppingCriterion,
    stopping_criterion_local_search: StoppingCriterion,    
) -> Optional[Solution]:

    best_sol = None # maybe add a greedy construction to be safe.
    best_f = math.inf

    step = 0
    while not stopping_criterion_outer(step, best_f):
        sol0 = randomized_constructor(I)        # random constructive phase
        sol1 = local_search(                   # local search refinement
            I,
            sol0,
            neighborhoods_cls,
            step_function,
            stopping_criterion_local_search,
        )

        f1 = objective(I, sol1)

        if f1 < best_f:
            best_f = f1
            best_sol = sol1
            
        step+=1

    return best_sol

File: cpp/core/src/grasp.cpp
#include <vector>
#include <algorithm>
#include <random>
#include <functional>
#include "solvers.hpp"
#include "utils.hpp"


// Solution GRASP::randomized_constructor_simple(
//     const Instance &I,
//     double a,
//     double alpha,
//     int max_tries)
// {
//     const int n  = I.n;
//     const int nK = I.nK;

//     // heuristic costs
//     std::vector<double> costs = utils::calc_my_metric(I, a);

//     // argsort(costs)
//     std::vector<int> perm(n);
//     std::iota(perm.begin(), perm.end(), 0);
//     std::sort(perm.begin(), perm.end(),
//               [&](int i, int j) { return costs[i] < costs[j]; });

//     Solution sol;
//     sol.routes.assign(nK, std::vector<int>{});

//     int served = 0;
//     std::vector<bool> used(n, false);

//     static thread_local std::mt19937 rng(std::random_device{}());

//     while (served < I.gamma)
//     {
//         // remaining requests (not yet used)
//         std::vector<int> remaining;
//         remaining.reserve(n);
//         for (int r : perm)
//             if (!used[r])
//                 remaining.push_back(r);

//         if (remaining.empty())
//             break;

//         int k = std::max(1, int(alpha * remaining.size()));
//         if (k > (int)remaining.size())
//             k = (int)remaining.size();

//         std::uniform_int_distribution<int> pick_rcl(0, k - 1);
//         int req = remaining[pick_rcl(rng)];
//         used[req] = true;

//         int pickup = 1 + req;
//         int drop   = 1 + I.n + req;
//         int dem    = I.demands[req];

//         bool inserted = false;

//         for (int t = 0; t < max_tries; ++t)
//         {
//             std::uniform_int_distribution<int> pick_route(0, nK - 1);
//             int vk = pick_route(rng);
//             auto &route = sol.routes[vk];
//             int m = (int)route.size();

//             // handle empty route explicitly
//             if (m == 0)
//             {
//                 if (dem <= I.C) {
//                     route.push_back(pickup);
//                     route.push_back(drop);
//                     inserted = true;
//                     served++;
//                     break;
//                 }
//                 // cannot serve on empty vehicle (capacity too small), try another vehicle/attempt
//                 continue;
//             }

//             // non-empty route: choose ip, jp safely
//             // ip in [0, m-1], jp in [ip+1, m]
//             std::uniform_int_distribution<int> pick_ip(0, m - 1);
//             int ip = pick_ip(rng);

//             std::uniform_int_distribution<int> pick_jp(ip + 1, m);
//             int jp = pick_jp(rng);

//             // build candidate route:
//             std::vector<int> new_r;
//             new_r.reserve(m + 2);

//             // [0, ip)
//             new_r.insert(new_r.end(), route.begin(), route.begin() + ip);
//             // pickup
//             new_r.push_back(pickup);
//             // [ip, jp)
//             new_r.insert(new_r.end(), route.begin() + ip, route.begin() + jp);
//             // drop
//             new_r.push_back(drop);
//             // [jp, m)
//             new_r.insert(new_r.end(), route.begin() + jp, route.end());

//             auto cargo = utils::calc_route_cargo(I, new_r);
//             bool ok = true;
//             for (int c : cargo)
//             {
//                 if (c < 0 || c > I.C)
//                 {
//                     ok = false;
//                     break;
//                 }
//             }

//             if (ok)
//             {
//                 route = std::move(new_r);
//                 inserted = true;
//                 served++;
//                 break;
//             }
//         }

//         // failed to insert this request in max_tries attempts → skip it
//         if (!inserted)
//             continue;
//     }

//     return sol;
// }


Solution GRASP::randomized_constructor_simple(
    const Instance &I,
    double a,
    double alpha)
{
    const int n  = I.n;
    const int nK = I.nK;

    std::vector<double> costs = utils::calc_my_metric(I, a);

    std::vector<int> perm(n);
    std::iota(perm.begin(), perm.end(), 0);
    std::sort(perm.begin(), perm.end(),
              [&](int i, int j){ return costs[i] < costs[j]; });

    Solution sol;
    sol.routes.assign(nK, {});

    std::vector<bool> used(n, false);
    int served = 0;

    static thread_local std::mt19937 rng(std::random_device{}());

    while (served < I.gamma)
    {
        // build remaining candidate list
        std::vector<int> cand;
        for (int r : perm)
            if (!used[r])
                cand.push_back(r);
        if (cand.empty()) break;

        // determine RCL by threshold
        double c_min = costs[cand[0]];
        double c_max = costs[cand.back()];
        double thresh = c_min + alpha * (c_max - c_min);

        std::vector<int> RCL;
        for (int r : cand)
            if (costs[r] <= thresh)
                RCL.push_back(r);

        if (RCL.empty())  // shouldn't happen, but safe
            RCL.push_back(cand[0]);

        // pick request from RCL at random
        std::uniform_int_distribution<int> pick_idx(0, RCL.size()-1);
        int req = RCL[pick_idx(rng)];
        used[req] = true;

        int pickup = 1 + req;
        int drop   = 1 + I.n + req;
        int dem    = I.demands[req];

        // find best feasible insertion for (pickup, drop)
        double best_inc = 1e18;
        int best_r = -1, best_ip = -1, best_jp = -1;

        for (int r = 0; r < nK; r++)
        {
            auto &route = sol.routes[r];
            int m = route.size();

            // attempt all insertions
            for (int ip = 0; ip <= m; ip++)
            {
                for (int jp = ip+1; jp <= m+1; jp++)
                {
                    std::vector<int> new_r;
                    new_r.reserve(m+2);

                    new_r.insert(new_r.end(), route.begin(), route.begin()+ip);
                    new_r.push_back(pickup);
                    new_r.insert(new_r.end(), route.begin()+ip, route.begin()+jp);
                    new_r.push_back(drop);
                    new_r.insert(new_r.end(), route.begin()+jp, route.end());

                    auto cargo = utils::calc_route_cargo(I, new_r);
                    bool ok=true;
                    for (int c : cargo)
                        if (c < 0 || c > I.C) { ok=false; break; }
                    if (!ok) continue;

                    // compute incremental cost
                    double inc = utils::route_distance(I, new_r)
                               - utils::route_distance(I, route);

                    if (inc < best_inc)
                    {
                        best_inc = inc;
                        best_r = r;
                        best_ip = ip;
                        best_jp = jp;
                    }
                }
            }
        }

        if (best_r == -1)
            continue; // cannot insert, skip

        // apply best insertion
        auto &route = sol.routes[best_r];
        std::vector<int> new_r;
        int m = route.size();
        new_r.reserve(m+2);
        new_r.insert(new_r.end(), route.begin(), route.begin()+best_ip);
        new_r.push_back(pickup);
        new_r.insert(new_r.end(), route.begin()+best_ip, route.begin()+best_jp);
        new_r.push_back(drop);
        new_r.insert(new_r.end(), route.begin()+best_jp, route.end());
        route = std::move(new_r);

        served++;
    }

    return sol;
}



Solution GRASP::grasp(
    const Instance &I,
    std::function<Solution(const Instance &)> randomized_constructor,
    const Neighborhood::NeighborhoodFactories &neighborhoods,
    StepFunction::Func step_function,
    StoppingCriterion &stopping_outer,
    StoppingCriterion &stopping_local)
{
    Solution best_sol;                         // final result
    double best_f = std::numeric_limits<double>::infinity();

    int step = 0;

    // reset local-search stop criterion every restart
    stopping_local.reset();
    stopping_outer.reset();
    
    while (!stopping_outer(step, best_f))
    {
        // === construct a new initial solution ===
        Solution sol0 = randomized_constructor(I);

        stopping_local.reset();
        Solution sol1 = LS::local_search(
            I,
            sol0,
            neighborhoods[step%neighborhoods.size()], //rotate neighborhood
            step_function,
            stopping_local);

        double f1 = utils::objective(I, sol1);

        if (f1 < best_f)
        {
            best_f = f1;
            best_sol = sol1;
        }

        step++;
    }

    return best_sol;
}
